lib/cretonne/src/ssa.rs

```diff
@@ -0,0 +1,653 @@
+    /// The SSA value is passed as an argument because it should be created with
+    /// `ir::DataFlowGraph::append_result`.
+    pub fn def_var(&mut self, var: Variable, val: Value, block: Block) {
+        if match self.variables.get_mut(var) {
+               Some(data) => {
+                   data.current_defs.insert(block, val);
+                   false
+               }
+               None => true,
+           } {
+            let mut defs = HashMap::new();
+            defs.insert(block, val);
+            self.variables
+                .insert(VariableData {
+                            current_defs: defs,
+                            variable: var,
+                        });
+        }
+    }
+
+    /// Declares a use of a variable in a given basic block.
+    /// Returns the SSA value corresponding to the current SSA definition of this variable.
+    pub fn use_var(&mut self,
+                   dfg: &mut DataFlowGraph,
+                   var: Variable,
+                   ty: Type,
+                   block: Block)
+                   -> Value {
+        // First we lookup for the current definition of the variable in this block
+        match self.variables.get(var) {
+            None => (),
+            Some(var_data) => {
+                match var_data.current_defs.get(&block) {
+                    None => (),
+                    Some(val) => {
+                        return *val;
+                    }
+                }
+            }
+        };
+        // At this point if we haven't returned it means that we have to search in the
+        // predecessors.
+        let (new_block, possible_val) = match self.blocks[block] {
+            BlockData::EbbHeader(ref preds, sealed, ebb, ref mut undef_variables) => {
+                // The block has multiple predecessors so we append an Ebb argument that
+                // will serve as a value.
+                // TODO: BUGFIX if the block is sealed then we have to append the branch argument
+                // to its predecessors
+                if sealed && preds.len() == 1 {
+                    (*preds.keys().next().unwrap(), None)
+                } else {
+                    let val = dfg.append_ebb_arg(ebb, ty);
+                    undef_variables.push((var, val));
+                    (block, Some((val, sealed, ebb)))
+                }
+            }
+            BlockData::EbbBody(pred, _) => (pred, None),
+        };
+        match possible_val {
+            // The block has a single predecessor, we look into it.
+            None => self.use_var(dfg, var, ty, new_block),
+            // The block has multiple predecessors, we register the ebb argument as the current
+            // definition for the variable.
+            Some((val, sealed, ebb)) => {
+                self.def_var(var, val, block);
+                if sealed {
+                    self.seal_ebb_header_block(ebb, dfg)
+                }
+                val
+            }
+        }
```

Comment:
Hrm. On the other hand, this may not be easy to do, because `def_var` needs a `&mut self`.